### PR TITLE
[23] Partition support for markdown comments #1542

### LINKED ISSUE #1542: [23] Partition support for markdown comments
Compiler support has now been added in jdt.core and code assist is planned via https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2744. My initial investigation shows that some work is needed in jdt.ui to make code completion requests to jdt.core APIs. I understand there's this notion of partition and we may need a new partition kind or treat /// same as Javadoc so that code completion requests are made and processed in the UI.

### PR DESCRIPTION
**Pull request by jarthana**
## What it does
Fixes #1542

## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java [hunk 1]
```diff
@@ -457,6 +461,9 @@ public IContentAssistant getContentAssistant(ISourceViewer sourceViewer) {
 			ContentAssistProcessor javadocProcessor= new JavadocCompletionProcessor(getEditor(), assistant);
 			assistant.setContentAssistProcessor(javadocProcessor, IJavaPartitions.JAVA_DOC);
 
+			ContentAssistProcessor markdownProcessor= new JavadocCompletionProcessor(getEditor(), assistant);
+			assistant.setContentAssistProcessor(javadocProcessor, IJavaPartitions.JAVA_MARKDOWN_COMMENT);
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java:465]: did you mean to pass markdownProcessor instead of javadocProcessor?

Comment (jarthana) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java]: Thanks, it's fixed now. Does it mean you have reviewed the patch?

Comment (stephan-herrmann) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java]: > Does it mean you have reviewed the patch?

I wouldn't call it a review. I was curious about what changes were needed, and stumbled upon this little glitch as a random finding :)

**Comment by stephan-herrmann** [comment]
It seems with pulling this change all markdown comments in the editor lose their coloring. Is this intended? Will you add new colors / preferences?

Or should be declare that javadoc == javadoc, independent of format? (At face value we're mainly replacing HTML with markdown, but it's still javadoc, right?). 

Or should we use colors to remind users of which format they are seeing, so they won't accidentally mix both?

**Comment by jarthana** [comment]
> It seems with pulling this change all markdown comments in the editor lose their coloring. Is this intended? Will you add new colors / preferences?

Not intentional. I will work on this.

> Or should be declare that javadoc == javadoc, independent of format? (At face value we're mainly replacing HTML with markdown, but it's still javadoc, right?).
> 
> Or should we use colors to remind users of which format they are seeing, so they won't accidentally mix both?

I would keep it simple and use the same color as Javadoc, at least for now.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java [hunk 2]
```diff
@@ -457,6 +465,9 @@ public IContentAssistant getContentAssistant(ISourceViewer sourceViewer) {
 			ContentAssistProcessor javadocProcessor= new JavadocCompletionProcessor(getEditor(), assistant);
 			assistant.setContentAssistProcessor(javadocProcessor, IJavaPartitions.JAVA_DOC);
 
+			ContentAssistProcessor markdownProcessor= new JavadocCompletionProcessor(getEditor(), assistant);
+			assistant.setContentAssistProcessor(markdownProcessor, IJavaPartitions.JAVA_MARKDOWN_COMMENT);
+
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java:470]: There's some inconsistency:
* the processor is registered for markdown partition
* the ctor of JavadocCompletionProcessor still associates it to the javadoc partition

When I tweak JavadocCompletionProcessor to pass the markdown partition into the super ctor, then no proposals will be shown in markdown any more.

IIUC, the reason is that plugin.xml does not associate any proposal computer with the markdown partition type.

Suggestion: add a parameter `partition` to the ctor of JavadocCompletionProcessor, then selectively add markdown partition in plugin.xml. This should give us control to enable desired processors while excluding, e.g., HTMLTagCompletionProposalComputer.

**Comment by stephan-herrmann** [comment]
Unsorted observations & ideas while playing with this PR and stepping through code:
* would it be interesting to define yet another partition type "codeblock within markdown"?
* I like the auto insertion of "///", how hard would it be to let this add as many blanks as the least indented line so far, minimally 1 (per convention)?
* folding completely hides the markdown region, whereas traditional javadoc will still show the first line
* seeing a change in `CodeTemplateSourceViewerConfiguration` (related to editing comment templates) lets me think we might want a full new set of comment templates in the future ...
* shift-hover over markdown link works, cool! :heavy_check_mark: 

Concerning completion:
* completion for `@param` works, both tag name and actual parameter name are proposed. :heavy_check_mark: 
* java templates are also proposed which is not useful
  * not a big issue, but at least priorities could be improved to show tag proposals above java templates
* after `@param` also HTML tags are proposed, which is not helpful
* I don't yet get the expected proposals after `[`
* After `[#` it proposes to insert `{@link #foo()}`, which is not what I want :), I want `[#foo()]`

In summary: I could not find any obvious bugs, so the PR can be seen as an incremental step forward. I don't think all of the above topics need to be dealt with while this is a preview feature.
* Perhaps the last item (completion after `[`) would be highest on my list, but then this needs to be addressed in jdt.core, not ui.
* On the UI side, perhaps folding (showing one line when folded) would be my first wish.
* For auto-insertion I'd suggest to just constantly insert one blank for now, to make it `/// `.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/AbstractFastJavaPartitionScanner.java [hunk 3]
```diff
@@ -326,9 +336,36 @@ public IToken nextToken() {
...
+				break;
+			case MARKDOWN_COMMENT:
+				switch (ch) {
+					case '\r':
+					case '\n':
+						return postFix(MARKDOWN_COMMENT);
+					case '/':
+						fTokenLength++;
+						fLast= SLASH_SLASH_SLASH;
+						break;
```
[the remark this answers is not in the document]
Comment (jarthana) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/AbstractFastJavaPartitionScanner.java:363]: You are right. I had thought it to be same as the asterisks in a regular Javadoc. But just now confirmed that the Javadoc is adding the fourth / as part of the text. This needs correction in all the places I touched, I think. Would prefer to take that as a follow-up commit.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java (hunk 2 above)
Comment (jarthana) [reply org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/JavaSourceViewerConfiguration.java:470]: In order to boost my confidence in this area (and to get this code to a build), I will take this up also as part of my next iteration.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/AbstractFastJavaPartitionScanner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephan-herrmann) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/AbstractFastJavaPartitionScanner.java:363]: fine by me.

**Comment by jarthana** [comment]
> 
>     * On the UI side, perhaps folding (showing one line when folded) would be my first wish.

The PR addresses this now.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/AbstractFastJavaPartitionScanner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jarthana) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/AbstractFastJavaPartitionScanner.java:363]: Created https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2929
